src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -0,0 +1,2164 @@
+            else if (internetStatus == Interop.WinHttp.WINHTTP_CALLBACK_STATUS_SENDING_REQUEST)
+            {
+                if (state.RequestMessage.RequestUri.Scheme != UriSchemeHttps)
+                {
+                    // Not SSL/TLS.
+                    return;
+                }
+
+                // Grab the channel binding token (CBT) information from the request handle and put it into
+                // the TransportContext object.
+                try
+                {
+                    state.TransportContext.SetChannelBinding(state.RequestHandle);
+                }
+                catch (Exception ex)
+                {
+                    Interop.WinHttp.WinHttpCloseHandle(handle);
+                    state.SavedException = ex;
+
+                    return;
+                }
+
+                if (state.ServerCertificateValidationCallback != null)
+                {
+                    IntPtr data = IntPtr.Zero;
+                    uint dataSize = (uint)IntPtr.Size;
+
+                    if (Interop.WinHttp.WinHttpQueryOption(
+                        state.RequestHandle,
+                        Interop.WinHttp.WINHTTP_OPTION_SERVER_CERT_CONTEXT,
+                        ref data,
+                        ref dataSize))
+                    {
+                        var serverCertificate = new X509Certificate2(data);
+                        Interop.Crypt32.CertFreeCertificateContext(data);
+
+                        X509Chain chain = null;
+                        SslPolicyErrors sslPolicyErrors;
+
+                        try
+                        {
+                            CertificateHelper.BuildChain(
+                                serverCertificate,
+                                state.RequestMessage.RequestUri.Host,
+                                state.CheckCertificateRevocationList,
+                                out chain,
+                                out sslPolicyErrors);
+
+                            var result = state.ServerCertificateValidationCallback(
+                                state.RequestMessage,
+                                serverCertificate,
+                                chain,
+                                sslPolicyErrors);
+                            if (!result)
+                            {
+                                Interop.WinHttp.WinHttpCloseHandle(handle);
+                                state.SavedException = 
+                                    WinHttpException.CreateExceptionUsingError((int)Interop.WinHttp.ERROR_WINHTTP_SECURE_FAILURE);
+                            }
+                        }
+                        catch (Exception ex)
+                        {
+                            Interop.WinHttp.WinHttpCloseHandle(handle);
+                            state.SavedException = ex;
+                        }
+                        finally
+                        {
+                            if (chain != null)
```

Comment:
While the `try..finally` here does address the marginal case where `out chain` is assigned before `BuildChain` otherwise throws an exception, this finally block suggests to me that it should instead be
```C#
CertificateHelper.BuildChain(...);

using (chain)
{
    ...
}
```